### PR TITLE
Minimal runtime wasm backend

### PR DESCRIPTION
**Pull request by juj**
This PR completes support for MINIMAL_RUNTIME in Wasm backend.

### CODE REVIEW

**Comment by sbc100** [comment]
Great!  Looks good to me.

File: emcc.py [hunk 1]
```diff
@@ -1577,7 +1577,7 @@ def is_supported_link_flag(f):
       shared.Settings.AUTO_JS_LIBRARIES = 0
 
       # In asm.js always use memory init file to get the best code size, other modes are not currently supported.
-      if not shared.Settings.WASM:
+      if not shared.Settings.WASM and not shared.Settings.WASM_BACKEND:
```
Comment (kripken) [inline emcc.py]: I think we still use a mem init file for wasm2js (wasm=0 with wasm backend). But maybe that's not handled in this code?

File: src/library.js [hunk 2]
```diff
@@ -302,13 +302,28 @@ LibraryManager.library = {
...
   getpagesize: function() {
     // int getpagesize(void);
+#if MINIMAL_RUNTIME // MINIMAL_RUNTIME does not have a global PAGE_SIZE runtime variable.
+#if WASM
+    return {{{ WASM_PAGE_SIZE }}};
+#else
+    return {{{ ASMJS_PAGE_SIZE }}};
+#endif
+#else
     return PAGE_SIZE;
```
Comment (kripken) [inline src/library.js]: How about leaving the minimal runtime code path here, for both runtimes? There's no benefit I can see to having both (basically, we can avoid using the global runtime variable in all cases, here).

File: src/library.js [hunk 3]
```diff
@@ -4548,9 +4567,9 @@ LibraryManager.library = {
 
   emscripten_with_builtin_malloc__deps: ['emscripten_builtin_malloc', 'emscripten_builtin_free', 'emscripten_builtin_memalign'],
   emscripten_with_builtin_malloc: function (func) {
-    var prev_malloc = _malloc;
-    var prev_memalign = _memalign;
-    var prev_free = _free;
+    var prev_malloc = typeof _malloc !== 'undefined' ? _malloc : undefined;
+    var prev_memalign = typeof _memalign !== 'undefined' ? _memalign : undefined;
+    var prev_free = typeof _free !== 'undefined' ? _free : undefined;
```
Comment (kripken) [inline src/library.js]: why are these changes needed?

File: src/library_syscall.js [hunk 4]
```diff
@@ -227,9 +227,15 @@ var SyscallsLibrary = {
...
     '$FS',
+#endif
+#if MINIMAL_RUNTIME // MINIMAL_RUNTIME does not have a global PAGE_SIZE runtime variable.
+    'getpagesize'
 #endif
   ],
   _emscripten_syscall_mmap2: function(addr, len, prot, flags, fd, off) {
+#if MINIMAL_RUNTIME // MINIMAL_RUNTIME does not have a global PAGE_SIZE runtime variable.
+    var PAGE_SIZE = _getpagesize();
+#endif
```
Comment (kripken) [inline src/library_syscall.js]: how about removing the ifdefs for minimal runtime, and always using `_getpagesize()` in both modes? that would keep things simpler overall, and I don't see much downside for the normal mode.

File: src/modules.js [hunk 5]
```diff
@@ -76,7 +76,9 @@ var LibraryManager = {
       libraries.push('library_exceptions_stub.js');
     }
 
-    if (!MINIMAL_RUNTIME) {
+    if (MINIMAL_RUNTIME) {
+      libraries.push('library_strings.js');
```
Comment (kripken) [inline src/modules.js]: it seems odd to have this included only in minimal runtime. perhaps a comment can clarify why this one is special?

File: src/library_syscall.js (hunk 4 above)
Comment (kripken) [reply src/library_syscall.js]: (again, my goal is to have as few code differences between the minimal and normal runtime, to reduce complexity and make it easier to migrate to it)

File: emcc.py (hunk 1 above)
Comment (juj) [reply emcc.py]: Yeah, this line only specifies memory init file for fastcomp asm.js builds.

File: src/library.js (hunk 2 above)
Comment (juj) [reply src/library.js]: That sounds good to me, but can that break some users that expect to be able to reference `PAGE_SIZE` variable in their library/--pre-js code that references the classic runtime? Perhaps that can be handled as a separate change?

File: src/library.js (hunk 3 above)
Comment (juj) [reply src/library.js]: In MINIMAL_RUNTIME code malloc/memalign/free might not be getting pulled in to the build, if not used in the program.

File: src/library_syscall.js (hunk 4 above)
Comment (juj) [reply src/library_syscall.js]: Sure, that works - I just did not want to regress code size for the classic runtime with my changes.

File: src/modules.js (hunk 5 above)
Comment (juj) [reply src/modules.js]: check

File: src/library.js (hunk 2 above)
Comment (kripken) [reply src/library.js]: Sorry if I wasn't clear - I'm not suggesting removing the `PAGE_SIZE` variable wherever it is declared for the normal runtime. I'm suggesting in this location here to not use that variable, and use the same code as the minimal runtime. It avoids an ifdef on the runtime, basically.

Comment (juj) [reply src/library.js]: Ok, I applied the change only for this location. I think that follows what you meant. Does that work?

Comment (kripken) [reply src/library.js]: lgtm with this change.

**Comment by kripken** [comment]
Perfect, thanks!

**Comment by juj** [comment]
In regular runtime, we have

```
var PAGE_SIZE = 16384;
```
I suppose that is for a historical reason? (or due to some performance tuning in mmap/node fopen?) I reverted the change to regular runtime, there are a lot of tests that verify that size.

**Comment by kripken** [comment]
I think this is a serious issue - this PR ended up creating a significant semantic difference between the two runtimes. I think we should either change the normal runtime to use wasm/asm.js page sizes instead of the old fixed one, or change the minimal runtime's page size to the old fixed value (which will be fine for code size).

Otherwise, I worry we'll end up with weird bugs in minimal runtime only, and have a hard time figuring them out.

(Are we perhaps mixing up multiple concepts of "page size" here?)

**Comment by juj** [comment]
> I think this is a serious issue

> (Are we perhaps mixing up multiple concepts of "page size" here?)

I don't know.

Let's revert if this is a problem. I don't have time to address this at the moment.
